microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshot-schema.md

```diff
@@ -0,0 +1,314 @@
+                    "internal",
+                    "hidden",
+                    "shortcut",
+                    "weak"
+                ],
+                "string_or_number",
+                "node"
+            ]
+        }
+    }
+}
+```
+
+### Nodes
+
+This `nodes` array, at the top-level of the `heapsnapshot` data, contains all the information needed to recreate the nodes of the graph.
+
+To parse this array, the following information is needed:
+
+* `snapshot.node_count` to know how many nodes there are.
+* `snapshot.meta.node_fields` to know how many fields each node has.
+
+Each node in the array is represented by a series of `snapshot.meta.node_fields.length` numbers. So the total number of elements in the `nodes` array is `snapshot.node_count` multiplied by `snapshot.meta.node_fields.length`.
+
+In order to recreate a node you need to read the numbers from the `nodes` array by groups of `snapshot.meta.node_fields.length`.
+
+Below is a code snippet that shows the node fields metadata and the data for the two first nodes in the graph.
+
+```json
+{
+    "snapshot": {
+        "meta": {
+            "node_fields": [
+                "type",
+                "name",
+                "id",
+                "self_size",
+                "edge_count",
+                "trace_node_id",
+                "detachedness"
+            ]
+            ...
+        }
+        ...
+    },
+    "nodes": [
+        9,1,1,0,10,0,0,
+        2,1,79,12,1,0,0,
+        ...
+    ]
+    ...
+}
+```
+
+| Index in node group | Name | Description |
+| --- | --- | --- |
+| `0` | `type` | The type of node. See [Node types](#node-types) to find out what are the possible types. |
+| `1` | `name` | The name of the node. This number corresponds to the index in the `strings` top-level array which you need to look-up to find the actual name. |
+| `2` | `id` | The node's unique ID. |
+| `3` | `self_size` | The node's size in bytes. |
+| `4` | `edge_count` | The number of edges connected to this node. |
+| `5` | `trace_node_id` | _Not documented yet_ |
+| `6` | `detachedness` | Whether this node can be reached from the `window` global object. `0` if it can be reached, `1` if it can't. |
+
+### Node types
+
+The first number found in the group of numbers for a node in the `nodes` array corresponds to its type. 
+
+This number is an index that can be used to lookup the type name in the `snapshot.meta.node_types[0]` array.
+
+| Node type | Description |
+| --- | --- |
+| Hidden | A V8 internal element that doesn't correspond directly to a user-controllable JavaScript object. In DevTools, these all show up under the category name **(system)**. Even though these objects are internal, they can be an important part of retainer paths. |
+| Object | Any user-defined object like `{ x: 2 }` or `new Foo(4)`. Contexts, which show up in DevTools as **system / Context** hold variables that had to be allocated on the heap because they're used by a nested function. |
```

Comment:
```suggestion
| Object | Any user-defined object, like `{ x: 2 }` or `new Foo(4)`. Contexts, which show up in DevTools as **system / Context** hold variables that had to be allocated on the heap because they're used by a nested function. |
```